Setting -Dwarnings breaks feature detection in build.rs

I'm not sure whether this is considered a bug due to custom `RUSTFLAGS`, but treating warnings as errors in CI doesn't sound uncommon.

**Version**

`cargo tree | grep tokio`
```
test-tokio-fail v0.1.0 (/tmp/test-tokio-fail)
└── tokio v1.24.1
```

**Platform**

`Linux <cut> 6.0.17-200.fc36.x86_64 #1 SMP PREEMPT_DYNAMIC Wed Jan 4 16:00:03 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux`

**Description**

Steps to reproduce:

```bash
rustup toolchain install 1.59.0
cargo new --bin test-tokio-fail
cd ./test-tokio-fail/
echo "tokio = \"=1.24.0\"" >> Cargo.toml # or 1.24.1, same issue

# We are treating warnings as errors in the CI
export RUSTFLAGS="-Dwarnings"
# Alternatively:
# export RUSTFLAGS="--cfg tokio_no_atomic_u64"

cargo +1.59.0 check
```

I get the following error:

```
   Compiling tokio v1.24.0
error[E0432]: unresolved import `crate::util::once_cell`
 --> /home/piodul/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.24.0/src/loom/std/atomic_u64_static_once_cell.rs:3:18
  |
3 | use crate::util::once_cell::OnceCell;
  |                  ^^^^^^^^^ could not find `once_cell` in `util`

For more information about this error, try `rustc --explain E0432`.
error: could not compile `tokio` due to previous error
```

My short investigation led me to the tokio's build.rs file. It looks like `-Dwarnings` affects the autocfg probe [here](https://github.com/tokio-rs/tokio/blob/c390a62387fe7346951c8bc57ea2761614b83e82/tokio/build.rs#L101), seen when running `cargo check` with high verbosity (`-vv`):

```
     Running `/tmp/test-tokio-fail/target/debug/build/tokio-0b45242da01fd5fb/build-script-build`
[tokio 1.24.0] error: unused import: `std::sync::atomic::AtomicU64 as _`
[tokio 1.24.0]  --> <anon>:3:9
[tokio 1.24.0]   |
[tokio 1.24.0] 3 |     use std::sync::atomic::AtomicU64 as _;
[tokio 1.24.0]   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[tokio 1.24.0]   |
[tokio 1.24.0]   = note: `-D unused-imports` implied by `-D warnings`
[tokio 1.24.0] 
[tokio 1.24.0] error: aborting due to previous error
[tokio 1.24.0] 
[tokio 1.24.0] cargo:rustc-cfg=tokio_no_target_has_atomic
[tokio 1.24.0] cargo:rustc-cfg=tokio_no_const_mutex_new
[tokio 1.24.0] cargo:rustc-cfg=tokio_no_atomic_u64
```

Because of this, `cargo:rustc-cfg=tokio_no_atomic_u64` is emitted in addition to others (I get the same effect with `RUSTFLAGS="--cfg tokio_no_atomic_u64"`). This cfg seems to be key to trigger the error.

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -307,6 +307,11 @@ jobs:
       - run: cargo test -Zbuild-std --target target-specs/i686-unknown-linux-gnu.json -p tokio --all-features
         env:
           RUSTFLAGS: --cfg tokio_unstable -Dwarnings --cfg tokio_no_atomic_u64
+      # https://github.com/tokio-rs/tokio/pull/5356
+      # https://github.com/tokio-rs/tokio/issues/5373
+      - run: cargo hack build -p tokio --feature-powerset --depth 2 -Z avoid-dev-deps --keep-going
+        env:
+          RUSTFLAGS: --cfg tokio_unstable -Dwarnings --cfg tokio_no_atomic_u64 --cfg tokio_no_const_mutex_new
 
   features:
     name: features

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -312,6 +312,9 @@ jobs:
       - run: cargo hack build -p tokio --feature-powerset --depth 2 -Z avoid-dev-deps --keep-going
         env:
           RUSTFLAGS: --cfg tokio_unstable -Dwarnings --cfg tokio_no_atomic_u64 --cfg tokio_no_const_mutex_new
+      - run: cargo hack build -p tokio --feature-powerset --depth 2 -Z avoid-dev-deps --keep-going
+        env:
+          RUSTFLAGS: --cfg tokio_unstable -Dwarnings --cfg tokio_no_atomic_u64
 
   features:
     name: features

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -300,6 +300,8 @@ jobs:
         with:
           toolchain: ${{ env.rust_nightly }}
           components: rust-src
+      - name: Install cargo-hack
+        uses: taiki-e/install-action@cargo-hack
       # Install linker and libraries for i686-unknown-linux-gnu
       - uses: taiki-e/setup-cross-toolchain-action@v1
         with:

--- a/tokio/build.rs
+++ b/tokio/build.rs
@@ -27,12 +27,13 @@ const CONST_MUTEX_NEW_PROBE: &str = r#"
 const TARGET_HAS_ATOMIC_PROBE: &str = r#"
 {
     #[cfg(target_has_atomic = "ptr")]
-    let _ = (); 
+    let _ = ();
 }
 "#;
 
 const TARGET_ATOMIC_U64_PROBE: &str = r#"
 {
+    #[allow(unused_imports)]
     use std::sync::atomic::AtomicU64 as _;
 }
 "#;

--- a/tokio/src/util/mod.rs
+++ b/tokio/src/util/mod.rs
@@ -6,7 +6,12 @@ cfg_io_driver! {
 #[cfg(feature = "rt")]
 pub(crate) mod atomic_cell;
 
-#[cfg(any(feature = "rt", feature = "signal", feature = "process"))]
+#[cfg(any(
+    feature = "rt",
+    feature = "signal",
+    feature = "process",
+    tokio_no_const_mutex_new,
+))]
 pub(crate) mod once_cell;
 
 #[cfg(any(

